packages/fx-core/src/component/generator/utils.ts

```diff
@@ -230,8 +234,69 @@ export function getSampleRelativePath(sampleName: string): string {
   return `${sampleConfig.baseFolderName}/${sampleName}/`;
 }
 
+export function getSampleUrl(sample: SampleInfo): string {
+  if (sample.url === sampleConfig.baseUrl) {
+    return `${sample.url}${sample.id}`;
+  } else {
+    return sample.url;
+  }
+}
+
 export function zipFolder(folderPath: string): AdmZip {
   const zip = new AdmZip();
   zip.addLocalFolder(folderPath);
   return zip;
 }
+
+export async function downloadDirectory(
+  sampleUrl: string,
+  dstPath: string,
+  concurrencyLimits = sampleConcurrencyLimits,
+  retryLimits = sampleDefaultRetryLimits
+): Promise<void> {
+  //step 1: get file info
+  const urlParserRegex = /https:\/\/github.com[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+  const [owner, repository, ref, sampleName] = urlParserRegex.exec(sampleUrl)!.slice(1);
+  const fileInfoUrl = `https://api.github.com/repos/${owner}/${repository}/git/trees/${ref}?recursive=1`;
+  const fileInfo = (
+    await sendRequestWithRetry(async () => {
+      return await axios.get(fileInfoUrl);
+    }, retryLimits)
+  ).data as any;
+
+  const filePrefixUrl = `https://raw.githubusercontent.com/${owner}/${repository}/${fileInfo.sha}/`;
+  const fileInfoTree = fileInfo.tree as any[];
+  const samplePaths = fileInfoTree
+    .filter((node) => node.path.startsWith(sampleName) && node.type !== "tree")
+    .map((node) => node.path);
+
+  //step 2: download files with limited concurrency
+  const downloadCallback = async (samplePath: string) => {
+    const file = await sendRequestWithRetry(async () => {
+      return await axios.get(filePrefixUrl + samplePath, { responseType: "arraybuffer" });
+    }, retryLimits);
+    const filePath = path.join(dstPath, samplePath.replace(sampleName, ""));
```

Comment:
why do we replace sample name with ""? Is it safer to use some method in path library to get relative path.

Reply:
Maybe you should try ``path.relative(samplePath, `/${sampleName}`)`` ?